server/app/controllers/CiviFormController.java

```diff
@@ -39,4 +40,19 @@ protected CompletableFuture<Void> checkProgramAdminAuthorization(
         .orElseThrow()
         .checkProgramAuthorization(programName, request);
   }
+
+  /** Checks that the profile is authorized to access the specified program. */
+  protected CompletableFuture<Void> checkProgramAuthorization(
+      ProfileUtils profileUtils,
+      VersionRepository versionRepository,
+      Http.Request request,
+      Long programId) {
+    return CompletableFuture.runAsync(
+        () -> {
+          if (versionRepository.isDraftProgram(programId)
+              && !profileUtils.currentUserProfile(request).orElseThrow().isCiviFormAdmin()) {
+            throw new SecurityException();
+          }
+        });
```

Comment:
**suggestion (blocking):** This blocks the a thread from the default thread pool on `versionRepository.isDraftProgram(programId)` to avoid this, add a new method to `versionRepository` that performs the IO using the [database connection threadpool](https://github.com/civiform/civiform/blob/main/server/app/repository/StoredFileRepository.java#L26):

```java
public CompletionStage<Boolean> isDraftProgramAsync(programId) {
  return supplyAsync(() -> isDraftProgram(programId), databaseExecutionContext);
}
```

The code here then consumes the new method and avoid `CompletableFuture.runAsync`

Reply:
Done